test/Dialect/FIRRTL/canonicalization.mlir

```diff
@@ -1249,4 +1249,21 @@ firrtl.module @mul_cst_prop3(%out_b: !firrtl.flip<sint<15>>) {
   firrtl.connect %out_b, %add : !firrtl.flip<sint<15>>, !firrtl.sint<15>
 }
 
+// We fold `validif` operations to their RHS, regardless of the LHS.
+// See https://github.com/llvm/circt/issues/839.
+// CHECK-LABEL: @elide_validif
+// CHECK-NEXT:      firrtl.connect %out0, %clock : !firrtl.flip<clock>, !firrtl.clock
+// CHECK-NEXT:      firrtl.connect %out1, %clock : !firrtl.flip<clock>, !firrtl.clock
+// CHECK-NEXT:  }
+firrtl.module @elide_validif(%clock: !firrtl.clock, %valid: !firrtl.uint<1>, %out0: !firrtl.flip<clock>, %out1: !firrtl.flip<clock>) {
+  %x = firrtl.wire  : !firrtl.clock
+  %y = firrtl.wire  : !firrtl.clock
+  %0 = firrtl.validif %valid, %clock : (!firrtl.uint<1>, !firrtl.clock) -> !firrtl.clock
+  %1 = firrtl.validif %valid, %clock : (!firrtl.uint<1>, !firrtl.clock) -> !firrtl.clock
+  firrtl.connect %x, %0 : !firrtl.clock, !firrtl.clock
+  firrtl.connect %y, %1 : !firrtl.clock, !firrtl.clock
+  firrtl.connect %out0, %x : !firrtl.flip<clock>, !firrtl.clock
+  firrtl.connect %out1, %y : !firrtl.flip<clock>, !firrtl.clock
+}
```

Comment:
Nit: Is there any reason to test two wires, as opposed to just one?

Reply:
Not really. That was some panicky attempt on my side to specifically check if we would get an `rtl.output` op with the same `%clock` SSA value. I'll remove that, adds nothing to the test but makes it twice as long.